diff --git a/bundle/nginx-1.27.1/src/http/ngx_http_upstream.h b/bundle/nginx-1.27.1/src/http/ngx_http_upstream.h
--- a/bundle/nginx-1.27.1/src/http/ngx_http_upstream.h
+++ b/bundle/nginx-1.27.1/src/http/ngx_http_upstream.h
@@ -31,6 +31,7 @@
 #define NGX_HTTP_UPSTREAM_FT_BUSY_LOCK       0x00001000
 #define NGX_HTTP_UPSTREAM_FT_MAX_WAITING     0x00002000
 #define NGX_HTTP_UPSTREAM_FT_NON_IDEMPOTENT  0x00004000
+#define NGX_HTTP_UPSTREAM_FT_HTTP_CUSTOM     0x00008000
 #define NGX_HTTP_UPSTREAM_FT_NOLIVE          0x40000000
 #define NGX_HTTP_UPSTREAM_FT_OFF             0x80000000
 
@@ -40,7 +41,8 @@
                                              |NGX_HTTP_UPSTREAM_FT_HTTP_504  \
                                              |NGX_HTTP_UPSTREAM_FT_HTTP_403  \
                                              |NGX_HTTP_UPSTREAM_FT_HTTP_404  \
-                                             |NGX_HTTP_UPSTREAM_FT_HTTP_429)
+                                             |NGX_HTTP_UPSTREAM_FT_HTTP_429  \
+                                             |NGX_HTTP_UPSTREAM_FT_HTTP_CUSTOM)
 
 #define NGX_HTTP_UPSTREAM_INVALID_HEADER     40
 
diff --git a/bundle/nginx-1.27.1/src/http/ngx_http_upstream.c b/bundle/nginx-1.27.1/src/http/ngx_http_upstream.c
--- a/bundle/nginx-1.27.1/src/http/ngx_http_upstream.c
+++ b/bundle/nginx-1.27.1/src/http/ngx_http_upstream.c
@@ -2679,10 +2679,24 @@
     ngx_msec_t                 timeout;
     ngx_uint_t                 status, mask;
     ngx_http_upstream_next_t  *un;
+#if (NGX_HTTP_LUA_KONG)
+    ngx_http_upstream_next_t   custom[2];
+#endif
 
     status = u->headers_in.status_n;
-
-    for (un = ngx_http_upstream_next_errors; un->status; un++) {
+    un = ngx_http_upstream_next_errors;
+
+#if (NGX_HTTP_LUA_KONG)
+    if (ngx_http_lua_kong_next_upstream_status(r, status)) {
+        custom[0].status = status;
+        custom[0].mask = NGX_HTTP_UPSTREAM_FT_HTTP_CUSTOM;
+        custom[1].status = 0;
+        custom[1].mask = 0;
+        un = custom;
+    }
+#endif
+
+    for ( /* void */; un->status; un++) {
 
         if (status != un->status) {
             continue;
@@ -4796,7 +4810,9 @@
         }
 
         if (ft_type == NGX_HTTP_UPSTREAM_FT_HTTP_403
-            || ft_type == NGX_HTTP_UPSTREAM_FT_HTTP_404)
+            || ft_type == NGX_HTTP_UPSTREAM_FT_HTTP_404
+            || (ft_type == NGX_HTTP_UPSTREAM_FT_HTTP_CUSTOM
+                && u->headers_in.status_n < 500))
         {
             state = NGX_PEER_NEXT;
 
@@ -4843,6 +4859,10 @@
 
     case NGX_HTTP_UPSTREAM_FT_HTTP_429:
         status = NGX_HTTP_TOO_MANY_REQUESTS;
+        break;
+
+    case NGX_HTTP_UPSTREAM_FT_HTTP_CUSTOM:
+        status = u->headers_in.status_n;
         break;
 
     /*
